// @activeledger/activelogger/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pino = require("pino");
const activeoptions_1 = require("@activeledger/activeoptions");
class ActiveLogger {
    static trace(p1, p2, args) {
        if (activeoptions_1.ActiveOptions.get("debug", false)) {
            if (typeof p1 == "object") {
                this.logger.trace(p1, p2, args || "");
            }
            else {
                this.logger.trace(p1, args || "");
            }
        }
    }
    trace(p1, p2, args) {
        ActiveLogger.trace(p1, p2);
    }
    static debug(p1, p2, args) {
        if (typeof p1 == "object") {
            this.logger.debug(p1, p2, args || "");
        }
        else {
            this.logger.debug(p1, args || "");
        }
    }
    debug(p1, p2, args) {
        ActiveLogger.debug(p1, p2);
    }
    static info(p1, p2, args) {
        if (typeof p1 == "object") {
            this.logger.info(p1, p2, args || "");
        }
        else {
            this.logger.info(p1, args || "");
        }
    }
    info(p1, p2, args) {
        ActiveLogger.info(p1, p2);
    }
    static warn(p1, p2, args) {
        if (typeof p1 == "object") {
            this.logger.warn(p1, p2, args || "");
        }
        else {
            this.logger.warn(p1, args || "");
        }
    }
    warn(p1, p2, args) {
        ActiveLogger.warn(p1, p2);
    }
    static error(p1, p2, args) {
        if (typeof p1 == "object") {
            this.logger.error(p1, p2, args || "");
        }
        else {
            this.logger.error(p1, args || "");
        }
    }
    error(p1, p2, args) {
        ActiveLogger.error(p1, p2);
    }
    static fatal(p1, p2, args) {
        if (typeof p1 == "object") {
            this.logger.fatal(p1, p2, args || "");
            return new Error(p2);
        }
        else {
            this.logger.fatal(p1, args || "");
            return new Error(p1);
        }
    }
    fatal(p1, p2, args) {
        ActiveLogger.fatal(p1, p2);
    }
}
ActiveLogger.logger = pino({
    name: "Activeledger",
    level: process.env.NODE_ENV == "production" ? "info" : "debug",
    prettyPrint: true
});
exports.ActiveLogger = ActiveLogger;
